engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableSelectBarrierTest.java

```diff
@@ -0,0 +1,683 @@
+            QueryTable.FORCE_PARALLEL_SELECT_AND_UPDATE = true;
+
+            final SelectColumn sa =
+                    SelectColumnFactory.getExpression("A=slow_a.getAndIncrementSlow(0, new int[]{ 0, 1 })");
+            final SelectColumn sb =
+                    SelectColumnFactory.getExpression("B=slow_a.getAndIncrementSlow(1, new int[]{ 1, 0 })");
+
+            slow_a.reset(0, -1);
+            System.out.println(Instant.now() + ": stateful");
+            final Table y = x.update(List.of(sa, sb));
+
+            final Table expected = y.updateView("A=i", "B=size + i");
+            assertTableEquals(expected, y);
+
+            slow_a.reset(0, 2);
+            System.out.println(Instant.now() + ": stateless");
+            final Table z = x.update(List.of(SelectColumn.ofStateless(sa), SelectColumn.ofStateless(sb)));
+
+            // these things should happen at the same time, so we should sum to the expected value;
+            // but everything should be mixed up
+            checkTotalSum(size, z);
+
+            // now check for swizzling
+            checkMixedColumns(z, true);
+
+            // and the column itself should not be ordered either
+            assertTrue(isOutOfOrder(z, "A"));
+            assertTrue(isOutOfOrder(z, "B"));
+
+            // now let's keep things stateless, but insert a barrier so that A and B do not intermix
+            slow_a.reset(0, 1);
+
+            System.out.println(Instant.now() + ": barrier");
+            final Table u = x.update(List.of(SelectColumn.ofStateless(sa).withBarriers(slow_a),
+                    SelectColumn.ofStateless(sb).respectsBarriers(slow_a)));
+
+            // check for swizzling
+            checkMixedColumns(u, false);
+
+            // we should have parallelized each column
+            assertTrue(isOutOfOrder(u, "A"));
+            assertTrue(isOutOfOrder(u, "B"));
+
+            // check the expected sums
+            checkIndividualSums(size, u);
+            System.out.println(Instant.now() + ": done");
+        }
+    }
+
+    // if you would like to convince yourself that we are reliably producing out-of-order conditions as appropriate
+    @Test
+    public void testRepeatedBarrierSelectable() {
+        for (int ii = 0; ii < REPEATS_FOR_CONFIDENCE; ++ii) {
+            System.out.println("Repetition " + ii);
+            try (final SafeCloseable ignored = LivenessScopeStack.open()) {
+                testBarrierSelectable();
+            }
+        }
+    }
+
+    @Test
+    public void testBarrierSelectable() {
+        QueryTable.FORCE_PARALLEL_SELECT_AND_UPDATE = true;
+
+        final int segments = 10;
+        final int size = Math.toIntExact(QueryTable.MINIMUM_PARALLEL_SELECT_ROWS * segments);
+        final SlowedAtomicInteger slow_a = new SlowedAtomicInteger(0, 2);
+        QueryScope.addParam("slow_a", slow_a);
+        QueryScope.addParam("size", size);
+        final Table x = TableTools.emptyTable(size)
```

Comment:
You could probably use `RawString` to let your formulas reference `QueryScope` params...